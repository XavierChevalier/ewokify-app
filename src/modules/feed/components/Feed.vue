<script setup lang="ts">
  import { mdiHeart } from '@mdi/js'
  import MDIcon from '@/components/icon/Icon.vue'
  import { computed } from 'vue'
  import { localeFormatDistance } from '@/tools/date/Date'

  const props = defineProps({
    author: {
      type: String,
      required: true,
    },
    content: {
      type: String,
      required: true,
    },
    numberOfLikes: {
      type: Number,
      required: true,
    },
    publicationDate: {
      type: Date,
      required: true,
    },
  })

  const publicationDateDistance = computed(() =>
    localeFormatDistance(props.publicationDate, new Date())
  )
</script>

<template>
  <span data-test="author">{{ author }}</span>
  <span data-test="publicationDate">{{ publicationDateDistance }}</span>
  <p data-test="content">{{ content }}</p>
  <MDIcon data-test="heart" :path="mdiHeart"></MDIcon>
  <span data-test="numberOfLikes">{{ numberOfLikes }}</span>
</template>
