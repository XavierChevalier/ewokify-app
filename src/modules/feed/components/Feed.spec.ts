import { mount } from '@vue/test-utils'
import Feed from './Feed.vue'
import { subDays } from 'date-fns'

describe('Feed', () => {
  function mountComponent(props: object = {}) {
    return mount(Feed, {
      props: {
        author: 'TheBaronless',
        content: 'Lorem ipsum',
        numberOfLikes: 10,
        publicationDate: new Date(),
        ...props,
      },
    })
  }

  it('should render correctly', () => {
    const wrapper = mountComponent()

    expect(wrapper.element).toMatchSnapshot()
  })

  it('should render the author name text', () => {
    const author = 'TheBaronless'
    const wrapper = mountComponent({ author })

    expect(wrapper.vm.author).toStrictEqual(author)
    expect(wrapper.find('[data-test="author"]').text()).toStrictEqual(author)
  })

  it('should render the feed content text', () => {
    const content = 'Lorem ipsum'
    const wrapper = mountComponent({ content })

    expect(wrapper.vm.content).toStrictEqual(content)
    expect(wrapper.find('[data-test="content"]').text()).toStrictEqual(content)
  })

  it('should render the publication date', () => {
    const publicationDate = subDays(new Date(), 2)
    const wrapper = mountComponent({ publicationDate })

    expect(wrapper.vm.publicationDate).toStrictEqual(publicationDate)
    expect(wrapper.find('[data-test="publicationDate"]').text()).toStrictEqual(
      '2 days ago'
    )
  })

  describe('likes', () => {
    it('should render the number of likes', () => {
      const numberOfLikes = 15
      const wrapper = mountComponent({ numberOfLikes })

      expect(wrapper.vm.numberOfLikes).toStrictEqual(numberOfLikes)
      expect(wrapper.find('[data-test="numberOfLikes"]').text()).toStrictEqual(
        numberOfLikes.toString()
      )
    })

    it('should render the heart icon', () => {
      const wrapper = mountComponent()

      expect(wrapper.find('[data-test="heart"]').exists()).toBeTruthy()
    })
  })
})
