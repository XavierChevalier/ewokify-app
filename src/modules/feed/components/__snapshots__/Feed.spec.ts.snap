// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Feed should render correctly 1`] = `
<div
  data-v-app=""
>
  
  <span
    data-test="author"
  >
    TheBaronless
  </span>
  <span
    data-test="publicationDate"
  >
    less than a minute ago
  </span>
  <p
    data-test="content"
  >
    Lorem ipsum
  </p>
  <svg
    data-test="heart"
    height="24"
    viewBox="0 0 24 24"
    width="24"
  >
    <path
      d="M12,21.35L10.55,20.03C5.4,15.36 2,12.27 2,8.5C2,5.41 4.42,3 7.5,3C9.24,3 10.91,3.81 12,5.08C13.09,3.81 14.76,3 16.5,3C19.58,3 22,5.41 22,8.5C22,12.27 18.6,15.36 13.45,20.03L12,21.35Z"
      data-test="iconPath"
    />
  </svg>
  <span
    data-test="numberOfLikes"
  >
    10
  </span>
  
</div>
`;
